src/main/scala/strawman/collection/immutable/ArraySeq.scala

```diff
@@ -0,0 +1,832 @@
+      val n = if (knownSize > 0) knownSize else xs.asInstanceOf[Iterable[A]].knownSize
+      val capacity = capacitate(n)
+      val array = scala.Array.ofDim[Any](capacity)
+      var i = 0
+      val it = xs.iterator()
+      while (i < n) {
+        array(i) = it.next()
+        i += 1
+      }
+      createPrimary(array, 0, n).toSecondary
+    case _ =>
+      val builder = ArraySeq.newBuilder[A]()
+      builder ++= xs
+      builder.result()
+  }
+
+  /**
+    * Used to wrap an existing array (without copying). Observe that later modifications to the specified array will show
+    * through to the returned instance.
+    * @param xs array to use for elements
+    * @param length the number of elements present in the array (defaults to -1 which will use the array length)
+    * @tparam A the element type to use
+    * @return a new instance using the specified array as element storage (or the empty ArraySeq if the array is empty).
+    */
+  private[strawman] def wrap[A](xs: Array[Any], length: Int = -1): ArraySeq[A] =
+    if (xs.length == 0) Empty
+    else createPrimary(xs, 0, if (length > -1) length else xs.length).toSecondary.withoutAutoShrinking
+
+  /** A class to build instances of `ArraySeq`.  This builder is reusable. */
+  final class Builder[A](var sizeHint: Int = -1, val shrinkThreshold: Int = DefaultAutoShrinkThreshold) extends ReusableBuilder[A, ArraySeq[A]] {
+    private var index = 0
+    private var elems: Array[Any] = allocate()
+
+    private def allocate(size: Int = sizeHint) = scala.Array.ofDim[Any](capacitate(size max 32))
+
+    override def clear() = elems = allocate()
+
+    override def sizeHint(size: Int) = sizeHint = size
+
+    override def result() =
+      if (index == 0) Empty
+      else ArraySeq.createPrimary[A](elems, 0, index).withAutoShrinking(shrinkThreshold)
+
+    override def add(elem: A) = {
+      if (elems.length == index) {
+        val array = allocate(index + 1)
+        java.lang.System.arraycopy(elems, 0, array, 0, index)
+        elems = array
+      }
+      elems(index) = elem
+      index += 1
+      this
+    }
+  }
+
+  def newBuilder[A](): Builder[A] = new Builder[A]()
+
+  def newSizedBuilder[A](sizeHint: Int): Builder[A] = new Builder[A](sizeHint)
+
+  def newSpecificBuilder[A](sizeHint: Int = -1, shrinkThreshold: Int = DefaultAutoShrinkThreshold): Builder[A] = new Builder[A](sizeHint = sizeHint, shrinkThreshold = shrinkThreshold)
+
+  override def empty[A <: Any]: ArraySeq[A] = Empty
+
+  @`inline`
+  private[ArraySeq] def indices(start: Int, stop: Int, capacity: Int): (Int, Int, Int, Int) = {
+    if (start < 0 && stop <= 0) (capacity + start, capacity + stop, 0, 0)
+    else if (start < 0) (capacity + start, capacity, 0, stop)
+    else if (stop < 0) (capacity + stop, capacity, 0, start)
+    else (0, 0, start, stop)
+  }
+
+  @`inline`
+  private[ArraySeq] final def distance(x: Int, y: Int): Int =
+    if (x < y) y - x
+    else x - y
+}
+
```

Comment:
The implementation looks good but I’m wondering what are the costs of having specialized implementations for almost all operation in terms of bytecode size and what are the performance boost?

Reply:
Any operation that delegates to `Iterator` has been overridden to avoid the creation of the iterator in order to get better performance.

The size of the generated classfiles are as follows (according to a manual summation):

Collection | Scala 2.12 | Scala 2.13 |  Dotty
---------- | ---------- | ---------- | -------
ArraySeq | 113 KB | 112 KB | 115 KB
Vector | 115 KB | 115 KB | 90 KB
List | 112 KB | 128 KB | 66 KB